snapcraft/internal/meta.py

```diff
@@ -215,19 +217,62 @@ def _wrap_exe(self, command, basename=None):
         return os.path.relpath(wrappath, self._snap_dir)
 
     def _wrap_apps(self, apps):
+        gui_dir = os.path.join(self.meta_dir, 'gui')
+        if not os.path.exists(gui_dir):
+            os.mkdir(gui_dir)
+        for f in os.listdir(gui_dir):
+            if os.path.splitext(f)[1] == '.desktop':
+                os.remove(os.path.join(gui_dir, f))
         for app in apps:
-            cmds = (k for k in ('command', 'stop-command') if k in apps[app])
-            for k in cmds:
-                try:
-                    apps[app][k] = self._wrap_exe(
-                        apps[app][k], '{}-{}'.format(k, app))
-                except CommandError as e:
-                    raise EnvironmentError(
-                        'The specified command {!r} defined in the app {!r} '
-                        'does not exist or is not executable'.format(
-                            str(e), app))
+            self._wrap_app(app, apps[app])
         return apps
 
+    def _wrap_app(self, name, app):
+        cmds = (k for k in ('command', 'stop-command') if k in app)
+        for k in cmds:
+            try:
+                app[k] = self._wrap_exe(app[k], '{}-{}'.format(k, name))
+            except CommandError as e:
+                raise EnvironmentError(
+                    'The specified command {!r} defined in the app {!r} '
+                    'does not exist or is not executable'.format(str(e), name))
+        if 'desktop' in app:
+            self._wrap_desktop_and_icon(name, app)
+
+    def _wrap_desktop_and_icon(self, name, app):
+        gui_dir = os.path.join(self.meta_dir, 'gui')
+        desktop_file = os.path.join(self._snap_dir, app['desktop'])
+        if not os.path.exists(desktop_file):
+            raise EnvironmentError(
+                'The specified desktop file {!r} defined in the app '
+                '{!r} does not exist'.format(desktop_file, name))
+        desktop_contents = configparser.ConfigParser(interpolation=None)
+        desktop_contents.optionxform = str
+        desktop_contents.read(desktop_file)
+        section = 'Desktop Entry'
+        if section not in desktop_contents.sections():
+            raise EnvironmentError(
+                'The specified desktop file {!r} is not a valid '
+                'desktop file'.format(desktop_file))
+        if 'Exec' not in desktop_contents[section]:
+            raise EnvironmentError(
+                'The specified desktop file {!r} is missing the '
+                '"Exec" key'.format(desktop_file))
+        # XXX: do we want to allow more parameters for Exec?
+        exec_value = '{}.{} %U'.format(self._config_data['name'], name)
+        desktop_contents[section]['Exec'] = exec_value
+        if 'Icon' in desktop_contents[section]:
+            icon = desktop_contents[section]['Icon']
```

Comment:
maybe fallback to the snap icon in meta/gui/icon if it exists, but this would be a different PR ;-)